nucypher/network/retrieval.py

```diff
@@ -0,0 +1,409 @@
+                            capsules: List[Capsule],
+                            cfrags: List[VerifiedCapsuleFrag],
+                            stamp: SignatureStamp,
+                            ) -> 'ReencryptionResponse':
+
+        # un-verify
+        cfrags = [CapsuleFrag.from_bytes(bytes(cfrag)) for cfrag in cfrags]
+
+        capsules_bytes = b''.join(bytes(capsule) for capsule in capsules)
+        cfrags_bytes = b''.join(bytes(cfrag) for cfrag in cfrags)
+        signature = stamp(capsules_bytes + cfrags_bytes)
+        return cls(cfrags, signature)
+
+    def __init__(self, cfrags: List[CapsuleFrag], signature: Signature):
+        self.cfrags = cfrags
+        self.signature = signature
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        signature, cfrags_bytes = signature_splitter(data, return_remainder=True)
+
+        # We would never send a request with no capsules, so there should be cfrags.
+        # The splitter would fail anyway, this just makes the error message more clear.
+        if not cfrags_bytes:
+            raise ValueError("ReencryptionResponse contains no cfrags")
+
+        cfrags = cfrag_splitter.repeat(cfrags_bytes)
+        return cls(cfrags, signature)
+
+    def __bytes__(self):
+        return bytes(self.signature) + b''.join(bytes(cfrag) for cfrag in self.cfrags)
+
+
+class RetrievalClient:
+    """
+    Capsule frag retrieval machinery shared between Bob and Porter.
+    """
+
+    def __init__(self, learner: Learner):
+        self._learner = learner
+        self.log = Logger(self.__class__.__name__)
+
+    def _ensure_ursula_availability(self, treasure_map: TreasureMap, timeout=10):
+        """
+        Make sure we know enough nodes from the treasure map to decrypt;
+        otherwise block and wait for them to come online.
+        """
+
+        # OK, so we're going to need to do some network activity for this retrieval.
+        # Let's make sure we've seeded.
+        if not self._learner.done_seeding:
+            self._learner.learn_from_teacher_node()
+
+        ursulas_in_map = treasure_map.destinations.keys()
+        all_known_ursulas = self._learner.known_nodes.addresses()
+
+        # Push all unknown Ursulas from the map in the queue for learning
+        unknown_ursulas = ursulas_in_map - all_known_ursulas
+        if unknown_ursulas:
+            self._learner.learn_about_specific_nodes(unknown_ursulas)
+
+        # If we know enough to decrypt, we can proceed.
+        known_ursulas = ursulas_in_map & all_known_ursulas
+        if len(known_ursulas) >= treasure_map.threshold:
+            return
+
+        # | <--- shares                                            ---> |
+        # | <--- threshold               ---> |
+        # | <--- known_ursulas ---> |         | <--- allow_missing ---> |
+        # |                         | <--- unknown_ursulas         ---> |
+        allow_missing = len(treasure_map.destinations) - treasure_map.threshold
```

Comment:
I think we might still be off here. 

The `allow_missing` value in this case represents "from the list of unknown ursulas, how many ursulas do we want to allow to not be learned about". That would be dependent on how many ursulas we know about already relative to the threshold.

If we didn't know about any ursulas, this calc would be correct, but if we already know about some ursulas, the value needs to be adjusted for that; basically, at this point in the code, we are short of the threshold by `(threshold - len(known_ursulas))` ursulas.

```suggestion
        allow_missing = len(unknown_ursulas) - (treasure_map.threshold - len(known_ursulas))
```

Alternatively, I think (?) you can skip all of this logic for known vs unknown ursulas and just pass the entire treasure map destinations since there is code in `block_until_specific_nodes_are_known()` to account for already known nodes anyway, and then use allow_missing=(len(destinations) - threshold)  i.e. something like

```python
self._learner.block_until_specific_nodes_are_known(
    addresses=treasure_map.destinations,
    timeout=timeout,
    allow_missing=(len(treasure_map.destinations) - treasure_map.threshold))
```

Reply:
> `len(unknown_ursulas) - (treasure_map.threshold - len(known_ursulas))`

Well, that amounts to the same thing, since `len(unknown_ursulas) + len(known_ursulas) == len(treasure_map.destinations)`

> I think (?) you can skip all of this logic for known vs unknown ursulas and just pass the entire treasure map destinations

Yes, that'll probably be simpler. The current calculation sort of evolved from nonzero redundancy which is now gone.

Reply:
> Well, that amounts to the same thing, since len(unknown_ursulas) + len(known_ursulas) == len(treasure_map.destinations)

You are right - that was less clear to me on first look 😊